### ISSUE TITLE
Does not work with Embroider

### ISSUE THREAD

**Issue by tansongyang** [true-myth/true-myth#67]
An Ember app with `true-myth` does not build with Embroider `v0.22.0`.

There was a recent [call to try out Embroider](https://the-emberjs-times.ongoodbits.com/2020/08/22/issue-161), so we did just that. One surprising result is that importing `true-myth` prevented our build from completing. We had to make a local copy of the library and changing all our imports.

**Comment by chriskrycho** [comment]
@tansongyang thank you for calling this out (and sorry for the delay in responding—only just saw this)!

Can you post the actual errors with Embroider? I expect this is *likely* a function of the way we're pre-building/bundling the various distributions (ESM, CJS, etc.), but more info to debug would be super helpful.

**Comment by wytlytningNZ** [comment]
👋 I am also experiencing this with Embroider.
This is the error I receive:
```
Build Error (PackagerRunner)

Module not found: Error: Can't resolve 'true-myth/maybe' in '/private/var/folders/0_/zsyyxymj6j7dq024g9vn02zh0000gn/T/embroider/18b9d4/routes/suburb/categories'
```

<details>
  <summary>And this is the full error</summary>
  
```
=================================================================================

ENV Summary:

  TIME: Fri Oct 09 2020 10:06:23 GMT+1100 (Australian Eastern Daylight Time)
  TITLE: ember
  ARGV:
  - /Users/me/.nvm/versions/node/v10.22.0/bin/node
  - /my-ember-app/node_modules/.bin/ember
  - s
  - --port=4200
  - --live-reload-port=0
  EXEC_PATH: /Users/me/.nvm/versions/node/v10.22.0/bin/node
  TMPDIR: /var/folders/0_/zsyyxymj6j7dq024g9vn02zh0000gn/T
  SHELL: /bin/zsh
  PATH:
  - /var/folders/0_/zsyyxymj6j7dq024g9vn02zh0000gn/T/yarn--1602198285443-0.7180013304810848
  - /my-ember-app/node_modules/.bin
  - /Users/me/.config/yarn/link/node_modules/.bin
  - /Users/me/.nvm/versions/node/v10.22.0/libexec/lib/node_modules/npm/bin/node-gyp-bin
  - /Users/me/.nvm/versions/node/v10.22.0/lib/node_modules/npm/bin/node-gyp-bin
  - /Users/me/.nvm/versions/node/v10.22.0/bin/node_modules/npm/bin/node-gyp-bin
  - /Users/me/.fastlane/bin
  - /Users/me/.local/bin
  - /Users/me/.rvm/gems/ruby-2.6.6/bin
  - /Users/me/.rvm/gems/ruby-2.6.6@global/bin
  - /Users/me/.rvm/rubies/ruby-2.6.6/bin
  - /Users/me/.nvm/versions/node/v10.22.0/bin
  - /usr/local/bin
  - /usr/bin
  - /bin
  - /usr/sbin
  - /sbin
  - /Applications/Postgres.app/Contents/Versions/latest/bin
  - /Users/me/.rvm/bin
  - /Users/me/Library/Android/sdk/platform-tools
  - /Users/me/.rvm/bin
  PLATFORM: darwin x64
  FREEMEM: 57831424
  TOTALMEM: 17179869184
  UPTIME: 2772751
  LOADAVG: 3.80810546875,2.80078125,2.3115234375
  CPUS:
  - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz - 2800
  - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz - 2800
  - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz - 2800
  - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz - 2800
  - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz - 2800
  - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz - 2800
  - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz - 2800
  - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz - 2800
  ENDIANNESS: LE
  VERSIONS:
  - ares: 1.15.0
  - brotli: 1.0.7
  - cldr: 35.1
  - http_parser: 2.9.3
  - icu: 64.2
  - modules: 64
  - napi: 6
  - nghttp2: 1.41.0
  - node: 10.22.0
  - openssl: 1.1.1g
  - tz: 2019c
  - unicode: 12.1
  - uv: 1.34.2
  - v8: 6.8.275.32-node.58
  - zlib: 1.2.11

ERROR Summary:

  - broccoliBuilderErrorStack: ModuleNotFoundError: Module not found: Error: Can't resolve 'true-myth/maybe' in '/private/var/folders/0_/zsyyxymj6j7dq024g9vn02zh0000gn/T/embroider/18b9d4/routes/suburb/categories'
    at factory.create (/my-ember-app/node_modules/webpack/lib/Compilation.js:925:10)
    at factory (/my-ember-app/node_modules/webpack/lib/NormalModuleFactory.js:401:22)
    at resolver (/my-ember-app/node_modules/webpack/lib/NormalModuleFactory.js:130:21)
    at asyncLib.parallel (/my-ember-app/node_modules/webpack/lib/NormalModuleFactory.js:224:22)
    at /my-ember-app/node_modules/webpack/node_modules/neo-async/async.js:2830:7
    at /my-ember-app/node_modules/webpack/node_modules/neo-async/async.js:6877:13
    at normalResolver.resolve (/my-ember-app/node_modules/webpack/lib/NormalModuleFactory.js:214:25)
    at doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:213:14)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/UnsafeCachePlugin.js:44:7)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:27:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:67:43)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn4 (<anonymous>:28:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/ModuleKindPlugin.js:30:40)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at args (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/forEachBail.js:30:14)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/UnsafeCachePlugin.js:44:7)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:27:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:67:43)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn5 (<anonymous>:16:1)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:27:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:67:43)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn1 (<anonymous>:16:1)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at fs.stat (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/DirectoryExistsPlugin.js:27:15)
    at process.nextTick (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:85:15)
    at process._tickCallback (internal/process/next_tick.js:61:11)

  - code: [undefined]
  - codeFrame: Module not found: Error: Can't resolve 'true-myth/maybe' in '/private/var/folders/0_/zsyyxymj6j7dq024g9vn02zh0000gn/T/embroider/18b9d4/routes/suburb/categories'
  - errorMessage: Module not found: Error: Can't resolve 'true-myth/maybe' in '/private/var/folders/0_/zsyyxymj6j7dq024g9vn02zh0000gn/T/embroider/18b9d4/routes/suburb/categories'
        at PackagerRunner (@embroider/webpack)
-~- created here: -~-
    at new Plugin (/my-ember-app/node_modules/@embroider/core/node_modules/broccoli-plugin/dist/index.js:45:31)
    at new PackagerRunner (/my-ember-app/node_modules/@embroider/core/src/to-broccoli-plugin.js:10:13)
    at Object.defaultPipeline (/my-ember-app/node_modules/@embroider/compat/src/default-pipeline.js:36:12)
    at module.exports (/my-ember-app/ember-cli-build.js:187:39)
    at Builder.readBuildFile (/my-ember-app/node_modules/ember-cli/lib/models/builder.js:49:14)
    at Builder.setupBroccoliBuilder (/my-ember-app/node_modules/ember-cli/lib/models/builder.js:63:22)
    at new Builder (/my-ember-app/node_modules/ember-cli/lib/models/builder.js:29:10)
    at ServeTask.run (/my-ember-app/node_modules/ember-cli/lib/tasks/serve.js:49:7)
    at Promise.resolve.then (/my-ember-app/node_modules/ember-cli/lib/models/command.js:238:24)
    at process._tickCallback (internal/process/next_tick.js:68:7)
    at Function.Module.runMain (internal/modules/cjs/loader.js:834:11)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)

-~- (end) -~-
  - errorType: Build Error
  - location:
    - column: [undefined]
    - file: [undefined]
    - line: [undefined]
    - treeDir: [undefined]
  - message: Module not found: Error: Can't resolve 'true-myth/maybe' in '/private/var/folders/0_/zsyyxymj6j7dq024g9vn02zh0000gn/T/embroider/18b9d4/routes/suburb/categories'
        at PackagerRunner (@embroider/webpack)
-~- created here: -~-
    at new Plugin (/my-ember-app/node_modules/@embroider/core/node_modules/broccoli-plugin/dist/index.js:45:31)
    at new PackagerRunner (/my-ember-app/node_modules/@embroider/core/src/to-broccoli-plugin.js:10:13)
    at Object.defaultPipeline (/my-ember-app/node_modules/@embroider/compat/src/default-pipeline.js:36:12)
    at module.exports (/my-ember-app/ember-cli-build.js:187:39)
    at Builder.readBuildFile (/my-ember-app/node_modules/ember-cli/lib/models/builder.js:49:14)
    at Builder.setupBroccoliBuilder (/my-ember-app/node_modules/ember-cli/lib/models/builder.js:63:22)
    at new Builder (/my-ember-app/node_modules/ember-cli/lib/models/builder.js:29:10)
    at ServeTask.run (/my-ember-app/node_modules/ember-cli/lib/tasks/serve.js:49:7)
    at Promise.resolve.then (/my-ember-app/node_modules/ember-cli/lib/models/command.js:238:24)
    at process._tickCallback (internal/process/next_tick.js:68:7)
    at Function.Module.runMain (internal/modules/cjs/loader.js:834:11)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)

-~- (end) -~-
  - name: Error
  - nodeAnnotation: @embroider/webpack
  - nodeName: PackagerRunner
  - originalErrorMessage: Module not found: Error: Can't resolve 'true-myth/maybe' in '/private/var/folders/0_/zsyyxymj6j7dq024g9vn02zh0000gn/T/embroider/18b9d4/routes/suburb/categories'
  - stack: ModuleNotFoundError: Module not found: Error: Can't resolve 'true-myth/maybe' in '/private/var/folders/0_/zsyyxymj6j7dq024g9vn02zh0000gn/T/embroider/18b9d4/routes/suburb/categories'
    at factory.create (/my-ember-app/node_modules/webpack/lib/Compilation.js:925:10)
    at factory (/my-ember-app/node_modules/webpack/lib/NormalModuleFactory.js:401:22)
    at resolver (/my-ember-app/node_modules/webpack/lib/NormalModuleFactory.js:130:21)
    at asyncLib.parallel (/my-ember-app/node_modules/webpack/lib/NormalModuleFactory.js:224:22)
    at /my-ember-app/node_modules/webpack/node_modules/neo-async/async.js:2830:7
    at /my-ember-app/node_modules/webpack/node_modules/neo-async/async.js:6877:13
    at normalResolver.resolve (/my-ember-app/node_modules/webpack/lib/NormalModuleFactory.js:214:25)
    at doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:213:14)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/UnsafeCachePlugin.js:44:7)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:27:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:67:43)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn4 (<anonymous>:28:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/ModuleKindPlugin.js:30:40)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at args (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/forEachBail.js:30:14)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/UnsafeCachePlugin.js:44:7)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:27:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:67:43)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn5 (<anonymous>:16:1)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:27:1)
    at resolver.doResolve (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:67:43)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn1 (<anonymous>:16:1)
    at hook.callAsync (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:285:5)
    at _fn0 (<anonymous>:15:1)
    at fs.stat (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/DirectoryExistsPlugin.js:27:15)
    at process.nextTick (/my-ember-app/node_modules/webpack/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:85:15)
    at process._tickCallback (internal/process/next_tick.js:61:11)


=================================================================================

```
</details>

**Comment by chriskrycho** [comment]
Ah, that’s very helpful, and I know what the problem is. Embroider follows Node’s module resolution, and the nested modules are not currently exposed in a way that resolves correctly for that (it would need to be in the root). I’ll think in the right solution here and cut a major with the fix in the near future!

**Comment by chriskrycho** [comment]
I've spent some time on this off and on over the last month, and I believe the *best* solution here will only be possible once TS supports the [Node 14 `exports` specifier][exports]. Combined with Node 14's stable support for ES modules, this package can switch to simply publishing ES modules with type definitions in a single location, and everything should “just work.”

The alternative, and what I think we probably need to do in the short term, is compile both the `.js` and `.d.ts` files to the root  of the repo and use the `files` key to name exactly and only those files to be published. I would likely also switch to publishing only ES modules in that approach. That *should* solve the problem identified here, and while it's less optimal than the `exports` mapping, it'll also “just work.”

@tansongyang @wytlytningNZ @bmakuh what are your thoughts?

[exports]: https://nodejs.org/dist/latest-v14.x/docs/api/packages.html#packages_package_entry_points

**Comment by tansongyang** [comment]
We're not planning to upgrade to Embroider in the near future. Mostly we just wanted to do an audit to see where we were at and what potential blockers might be. So I'm fine with waiting.

**Comment by chriskrycho** [comment]
Very good. Thanks for doing the investigation and giving the report early. Makes it much easier for me to tackle before it's urgent!

**Comment by chriskrycho** [comment]
Update: my current thinking here (pending further input from @bmakuh) is to cut a release that drops support for Node 10—possibly a bit earlier than Node 10 itself stops being supported, but we'll see when we get to it, since Node 12 LTS, 14 LTS, and 15+ all support the new `exports` mapping. That *should* allow us to dramatically simplify things by just publishing a single set of type definitions and a single set of ESM exports, which will make the build configuration way simpler *and* make everything "just work" for all downstream consumers. I'll update this once I have a PR in place which does that!

**Comment by bmakuh** [comment]
I think that's fine; if someone is stuck on node 10 they can stick with an old version of TM.

**Comment by chriskrycho** [comment]
I *believe* this should be resolved by the addition of export maps (#126) in the forthcoming 5.0 release (which should be out in the next few hours if I'm able to make sure everything is good)!

**Comment by chriskrycho** [comment]
Release as part of v5. @tansongyang @kiwi-josh would you good folks mind confirming this works as expected with Embroider (or even just a more basic ember-auto-import) usage now? I tested on my end and it *should*, since it no longer has any Ember-specific behavior to it at all, but it would be good to know if it *doesn't* for some reason!
